src/main/java/com/librato/disco/Level2StateCache.java

```diff
@@ -0,0 +1,132 @@
+package com.librato.disco;
+
+import org.apache.curator.framework.recipes.cache.ChildData;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class Level2StateCache implements IStateCache {
+    private static final Logger log = LoggerFactory.getLogger(Level2StateCache.class);
+    private final IStateCache delegate;
+    private final Level2CacheStrategy strat;
+    private final AtomicBoolean promoted = new AtomicBoolean();
+    private final ConcurrentMap<ChildData, Long> cache = new ConcurrentHashMap<>();
+    private final AtomicReference<List<ChildData>> promotedData = new AtomicReference<>();
+
+    /**
+     * Constructor.
+     *
+     * @param delegate the cache that will populate the l2 cache
+     * @param strat    determines when to promote demote
+     */
+    public Level2StateCache(IStateCache delegate, Level2CacheStrategy strat) {
+        this.delegate = delegate;
+        this.strat = strat;
+    }
+
+    @Override
+    public List<ChildData> getCurrentData() {
+        List<ChildData> data = delegate.getCurrentData();
+        if (strat == null) {
+            // cache is disbled if strat is null
+            return data;
+        }
+
+        prune();
+        updateCache(data);
+
+        // signal to the strat that we have delegate data
+        strat.accept(data);
+
+        List<ChildData> promotedData = getPromotedData();
+        if (promotedData != null && !promotedData.isEmpty()) {
+            return promotedData;
+        }
+
+        // by default return the delegate data
+        return data;
+    }
+
+    /**
+     * A synchronized method which first asks the strategy whether or not
+     * the cache should be in a promoted state.
+     *
+     * If the strategy recommends promotion then it will attempt to set the
+     * state of the cache to promoted and will set a reference to the
+     * data frozen at the time of promotion. If the strategy recommends promotion
+     * it will return the current frozen data.
+     *
+     * If the strategy does not recommend promotion then it will return null
+     * to signify that it should use the upstream delegate cached data.
+     */
+    private synchronized List<ChildData> getPromotedData() {
+        if (strat.promote()) {
+            if (this.promoted.compareAndSet(false, true)) {
+                ArrayList<ChildData> newPromotion = new ArrayList<>(cache.keySet());
+                log.info("Promoting L2 cache using {} promoted child data nodes", newPromotion.size());
+                promotedData.set(newPromotion);
```

Comment:
We may want to record a timestamp when we promote so that we could alert when a promotion lasts longer than a particular time.